File: packages/canvas/src-native/canvas-android/canvas/src/main/java/com/github/triniwiz/canvas/extensions/EXT_disjoint_timer_query.java
package com.github.triniwiz.canvas.extensions;

import android.opengl.GLES20;
import android.opengl.GLES30;
import android.os.Build;

import androidx.annotation.RequiresApi;

import com.github.triniwiz.canvas.CanvasView;
import com.github.triniwiz.canvas.Constants;

import java.util.concurrent.CountDownLatch;

/**
 * Created by triniwiz on 5/1/20
 */
@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
public class EXT_disjoint_timer_query {
    public static int QUERY_COUNTER_BITS_EXT = Constants.GL_QUERY_COUNTER_BITS_EXT;
    public static int CURRENT_QUERY_EXT = Constants.GL_CURRENT_QUERY_EXT;
    public static int QUERY_RESULT_EXT = Constants.GL_QUERY_RESULT_EXT;
    public static int QUERY_RESULT_AVAILABLE_EXT = Constants.GL_QUERY_RESULT_AVAILABLE_EXT;
    public static int TIME_ELAPSED_EXT = Constants.GL_TIME_ELAPSED_EXT;
    public static int TIMESTAMP_EXT = Constants.GL_TIMESTAMP_EXT;
    public static int GPU_DISJOINT_EXT = Constants.GL_GPU_DISJOINT_EXT;
    CanvasView canvas;

    public EXT_disjoint_timer_query(CanvasView canvas) {
        this.canvas = canvas;
    }

    public int createQueryEXT() {
        final CountDownLatch lock = new CountDownLatch(1);
        final int[] query = new int[1];
        canvas.queueEvent(new Runnable() {
            @Override
            public void run() {
                GLES30.glGenQueries(1, query, 0);
                lock.countDown();
            }
        });
        try {
            lock.await();
        } catch (InterruptedException ignored) {
        }
        return query[0];
    }

    public void deleteQueryEXT(int query) {
        final CountDownLatch lock = new CountDownLatch(1);
        final int[] id = {query};
        canvas.queueEvent(new Runnable() {
            @Override
            public void run() {
                GLES30.glDeleteQueries(1, id, 0);
                lock.countDown();
            }
        });
        try {
            lock.await();
        } catch (InterruptedException ignored) {
        }
    }

    public boolean isQueryEXT(final int query) {
        final CountDownLatch lock = new CountDownLatch(1);
        final boolean[] value = new boolean[1];
        canvas.queueEvent(new Runnable() {
            @Override
            public void run() {
                value[0] = GLES30.glIsQuery(query);
                lock.countDown();
            }
        });
        try {
            lock.await();
        } catch (InterruptedException ignored) {
        }
        return value[0];
    }

    public void beginQueryEXT(final int target, final int query) {
        final CountDownLatch lock = new CountDownLatch(1);
        canvas.queueEvent(new Runnable() {
            @Override
            public void run() {
                GLES30.glBeginQuery(target, query);
                lock.countDown();
            }
        });
        try {
            lock.await();
        } catch (InterruptedException ignored) {
        }
    }

    public void endQueryEXT(final int target) {
        final CountDownLatch lock = new CountDownLatch(1);
        canvas.queueEvent(new Runnable() {
            @Override
            public void run() {
                GLES30.glEndQuery(target);
                lock.countDown();
            }
        });
        try {
            lock.await();
        } catch (InterruptedException ignored) {
        }
    }

    public void queryCounterEXT(int query, int target) {
        // NOOP
    }

    public int getQueryEXT(final int target, final int pname) {
        final CountDownLatch lock = new CountDownLatch(1);
        final int[] query = new int[1];
        canvas.queueEvent(new Runnable() {
            @Override
            public void run() {
                GLES30.glGetQueryiv(target, pname, query, 0);
                lock.countDown();
            }
        });
        try {
            lock.await();
        } catch (InterruptedException ignored) {
        }
        return query[0];
    }

    public Object getQueryObjectEXT(final int query, final int pname) {
        final CountDownLatch lock = new CountDownLatch(1);
        final int[] value = new int[1];
        canvas.queueEvent(new Runnable() {
            @Override
            public void run() {
                GLES30.glGetQueryObjectuiv(query, pname, value, 0);
                lock.countDown();
            }
        });
        try {
            lock.await();
        } catch (InterruptedException ignored) {
        }
        if(pname ==  QUERY_RESULT_AVAILABLE_EXT){
            return value[0] == GLES20.GL_TRUE;
        }
        return value[0];
    }
}

File: packages/canvas/src-native/canvas-android/canvas/src/main/java/org/nativescript/canvas/extensions/EXT_disjoint_timer_query.kt
package org.nativescript.canvas.extensions

import android.opengl.GLES20
import android.opengl.GLES30
import android.os.Build
import androidx.annotation.RequiresApi
import org.nativescript.canvas.Constants
import org.nativescript.canvas.TNSCanvas
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Created by triniwiz on 5/1/20
 */
@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
class EXT_disjoint_timer_query(var canvas: TNSCanvas) {
	fun createQueryEXT(): Int {
		val query = IntArray(1)
		GLES30.glGenQueries(1, query, 0)
		return query[0]
	}

	fun deleteQueryEXT(query: Int) {
		val id = intArrayOf(query)
		GLES30.glDeleteQueries(1, id, 0)
	}

	fun isQueryEXT(query: Int): Boolean {
		return GLES30.glIsQuery(query)
	}

	fun beginQueryEXT(target: Int, query: Int) {
		GLES30.glBeginQuery(target, query)
	}

	fun endQueryEXT(target: Int) {
		GLES30.glEndQuery(target)
	}

	fun queryCounterEXT(query: Int, target: Int) {
		// NOOP
	}

	fun getQueryEXT(target: Int, pname: Int): Int {
		val query = IntArray(1)
		GLES30.glGetQueryiv(target, pname, query, 0)
		return query[0]
	}

	fun getQueryObjectEXT(query: Int, pname: Int): Any {
		val value = IntArray(1)
		GLES30.glGetQueryObjectuiv(query, pname, value, 0)
		return if (pname == QUERY_RESULT_AVAILABLE_EXT) {
			value[0] == GLES20.GL_TRUE
		} else value[0]
	}

	companion object {
		var QUERY_COUNTER_BITS_EXT = Constants.GL_QUERY_COUNTER_BITS_EXT
		var CURRENT_QUERY_EXT = Constants.GL_CURRENT_QUERY_EXT
		var QUERY_RESULT_EXT = Constants.GL_QUERY_RESULT_EXT
		var QUERY_RESULT_AVAILABLE_EXT = Constants.GL_QUERY_RESULT_AVAILABLE_EXT
		var TIME_ELAPSED_EXT = Constants.GL_TIME_ELAPSED_EXT
		var TIMESTAMP_EXT = Constants.GL_TIMESTAMP_EXT
		var GPU_DISJOINT_EXT = Constants.GL_GPU_DISJOINT_EXT
	}
}
